src/ucp/core/ucp_request.inl

```diff
@@ -165,6 +165,124 @@ void ucp_request_recv_generic_dt_finish(ucp_request_t *req)
 }
 
 static UCS_F_ALWAYS_INLINE void
+ucp_request_send_state_init(ucp_request_t *req, size_t dt_count)
+{
+    ucp_dt_generic_t *dt_gen;
+    void             *state_gen;
+
+    VALGRIND_MAKE_MEM_UNDEFINED(&req->send.uct_comp.count,
+                                sizeof(req->send.uct_comp.count));
+    VALGRIND_MAKE_MEM_UNDEFINED(&req->send.state.offset,
+                                sizeof(req->send.state.offset));
+
+    req->send.uct_comp.func = NULL;
+
+    switch (req->send.datatype & UCP_DATATYPE_CLASS_MASK) {
+    case UCP_DATATYPE_CONTIG:
+        return;
+    case UCP_DATATYPE_IOV:
+        req->send.state.dt.iov.iovcnt_offset = 0;
+        req->send.state.dt.iov.iov_offset    = 0;
+        req->send.state.dt.iov.iovcnt        = dt_count;
+        return;
+    case UCP_DATATYPE_GENERIC:
+        dt_gen    = ucp_dt_generic(req->send.datatype);
+        state_gen = dt_gen->ops.start_pack(dt_gen->context, req->send.buffer,
+                                           dt_count);
+        req->send.state.dt.generic.state = state_gen;
+        return;
+    default:
+        ucs_error("Invalid data type");
+    }
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_request_send_state_reset(ucp_request_t *req,
+                             uct_completion_callback_t comp_cb, unsigned proto)
+{
+    switch (proto) {
+    case UCP_REQUEST_SEND_PROTO_RMA:
+        ucs_assert(UCP_DT_IS_CONTIG(req->send.datatype));
+        /* Fall through */
+    case UCP_REQUEST_SEND_PROTO_RNDV_GET:
+        if (UCP_DT_IS_CONTIG(req->send.datatype)) {
+            req->send.state.dt.contig.memh = UCT_MEM_HANDLE_NULL;
+        }
+        /* Fall through */
+    case UCP_REQUEST_SEND_PROTO_ZCOPY_AM:
+        req->send.uct_comp.func       = comp_cb;
+        req->send.uct_comp.count      = 0;
+        /* Fall through */
+    case UCP_REQUEST_SEND_PROTO_BCOPY_AM:
+        req->send.state.offset           = 0;
+        break;
+    default:
+        ucs_fatal("unknown protocol");
+    }
+
+    /* offset is not used for RMA */
+    ucs_assert((proto == UCP_REQUEST_SEND_PROTO_RMA) ||
+               (req->send.state.offset <= req->send.length));
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_request_send_state_advance(ucp_request_t *req,
+                               const ucp_dt_state_t *dt_state, unsigned proto,
```

Comment:
code style - pls arrange parameters so single line will be as long as possible (but up to 80)
also rename dt_state to "new_dt_state" and add comment to explain this function usage

Reply:
the same: > 80. i will rename `dt_state` arg and add comment